src/section/mod.rs

```diff
@@ -8,27 +8,395 @@
 
 mod elders_info;
 mod member_info;
-mod network_stats;
-mod prefix_map;
 mod section_keys;
-mod section_map;
 mod section_peers;
 mod section_proof_chain;
-mod section_update_barrier;
-mod shared_state;
 
 #[cfg(test)]
 pub(crate) use self::elders_info::gen_elders_info;
+pub(crate) use self::section_peers::SectionPeers;
 pub use self::{
     elders_info::{majority_count, EldersInfo},
     member_info::{MemberInfo, PeerState, MIN_AGE},
-    network_stats::NetworkStats,
     section_keys::{SectionKeyShare, SectionKeysProvider},
-    section_map::SectionMap,
-    section_peers::SectionPeers,
     section_proof_chain::{ExtendError, SectionProofChain, TrustStatus},
 };
-pub(crate) use self::{
-    section_update_barrier::SectionUpdateBarrier,
-    shared_state::{SharedState, UpdateSectionKnowledgeAction},
+
+use crate::{
+    consensus::Proven,
+    error::{Error, Result},
+    peer::Peer,
+    rng, NetworkParams,
+};
+use bls_signature_aggregator::Proof;
+use serde::Serialize;
+use std::{
+    cmp::Ordering,
+    collections::{BTreeMap, BTreeSet},
+    convert::TryInto,
+    iter,
+    net::SocketAddr,
 };
+use xor_name::{Prefix, XorName};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub(crate) struct Section {
+    members: SectionPeers,
+    elders_info: Proven<EldersInfo>,
+    chain: SectionProofChain,
+}
+
+impl Section {
+    /// Creates a minimal `Section` initially containing only info about our elders
+    /// (`elders_info`).
+    ///
+    /// # Panics
+    ///
+    /// Panics if the key used to sign `elders_info` is not present in `section_chain`.
+    pub fn new(chain: SectionProofChain, elders_info: Proven<EldersInfo>) -> Self {
+        assert!(chain.has_key(&elders_info.proof.public_key));
+
+        Self {
+            elders_info,
+            chain,
+            members: SectionPeers::default(),
```

Comment:
would like to confirm that shall the `members` contains the `elders` ?

Reply:
No. Non elder nodes have `members` empty and elders populate it separately. We can't populate it given `EldersInfo` alone, because we also need the proofs.